packages/flutter_tools/lib/src/base/logger.dart

```diff
@@ -395,123 +422,154 @@ enum _LogType { error, status, trace }
     assert(_isStarted);
+    assert(stopwatch.isRunning);
+    stopwatch.stop();
     _isStarted = false;
     if (onFinish != null)
       onFinish();
   }
 }
 
+/// A [SilentStatus] shows nothing.
+class SilentStatus extends Status {
+  SilentStatus({
+    @required Duration timeout,
+    VoidCallback onFinish,
+  }) : super(timeout: timeout, onFinish: onFinish);
+}
+
 /// An [AnsiSpinner] is a simple animation that does nothing but implement a
-/// ASCII/Unicode spinner. When stopped or canceled, the animation erases
-/// itself.
+/// terminal spinner. When stopped or canceled, the animation erases itself.
 class AnsiSpinner extends Status {
-  AnsiSpinner({VoidCallback onFinish}) : super(onFinish: onFinish);
+  AnsiSpinner({
+    @required Duration timeout,
+    VoidCallback onFinish,
+  }) : super(timeout: timeout, onFinish: onFinish);
 
   int ticks = 0;
   Timer timer;
 
+  bool _showingSlowWarning = false;
+
   // Windows console font has a limited set of Unicode characters.
   List<String> get _animation => platform.isWindows
       ? <String>[r'-', r'\', r'|', r'/']
       : <String>['⣾', '⣽', '⣻', '⢿', '⡿', '⣟', '⣯', '⣷'];
 
-  String get _backspace => '\b' * _animation[0].length;
-  String get _clear => ' ' *  _animation[0].length;
+  static const String _slowWarning = ' (This is taking an unexpectedly long time.)';
 
-  void _callback(Timer timer) {
-    stdout.write('$_backspace${_animation[ticks++ % _animation.length]}');
-  }
+  String get _currentAnimationFrame => _animation[ticks % _animation.length];
+  int get _currentLength => _currentAnimationFrame.length + (_showingSlowWarning ? _slowWarning.length : 0);
+  String get _backspace => '\b' * _currentLength;
+  String get _clear => ' ' *  _currentLength;
 
   @override
   void start() {
     super.start();
     assert(timer == null);
-    stdout.write(' ');
+    stdout.write(_clear); // for _callback to backspace over
     timer = Timer.periodic(const Duration(milliseconds: 100), _callback);
     _callback(timer);
   }
 
   @override
-  void stop() {
+  void finish() {
     assert(timer.isActive);
     timer.cancel();
     stdout.write('$_backspace$_clear$_backspace');
-    super.stop();
+    super.finish();
   }
 
-  @override
-  void cancel() {
-    assert(timer.isActive);
-    timer.cancel();
-    stdout.write('$_backspace$_clear$_backspace');
-    super.cancel();
+  void _callback(Timer timer) {
+    stdout.write('$_backspace');
+    ticks += 1;
+    stdout.write('$_currentAnimationFrame');
+    if (seemsSlow) {
+      stdout.write(_slowWarning);
+      _showingSlowWarning = true;
+    }
   }
 }
 
 /// Constructor writes [message] to [stdout] with padding, then starts as an
-/// [AnsiSpinner].  On [cancel] or [stop], will call [onFinish].
-/// On [stop], will additionally print out summary information in
-/// milliseconds if [expectSlowOperation] is false, as seconds otherwise.
+/// [AnsiSpinner]. On [cancel] or [stop], will call [onFinish]. On [stop], will
```

Comment:
Cognitive overhead == high: this _is_ an `AnsiSpinner`, and it also _starts_ an `AnsiSpinner`?

Reply:
tried to clarify